ecosystem/op-deployer.md

```diff
@@ -0,0 +1,163 @@
+be stored by default within the intent's TOML file, however the implementation of the "state store" will be
+pluggable in order to support use cases like `opc` which store the deployment state in a database. Example output from a
+stage is below:
+
+```toml
+[intent]
+# Intent data is same as above and elided
+
+[state.chains.901]
+# Ownership addresses derived from a mnemonic
+proxyAdminOwner = "0xe59a881b2626f948f56f509f180c32428585629a"
+finalSystemOwner = "0xb9cdf788704088a4c0191d045c151fcbe2db14a4"
+baseFeeVaultRecipient = "0xbc4a9110dad00b4d9fb61743598848ddda6eeb03"
+l1FeeVaultRecipient = "0x6b0c2542fa2cadced5c7f64ef6fb9ebbce7630ff"
+sequencerFeeVaultRecipient = "0xb4e5b724bbc54c95d292613d956871281120ead6"
+
+# Address of OPSM so that other addresses can be retrieved from on-chain data
+opStackManagerAddress = "0x79c6c6b1844e3db7c30107f189cfb095bd2c4b5d"
+
+# Genesis data
+genesis = "base64://abcd..."
+genesisHash = "0x1234..."
+genesisTime = 1234567890
+
+[state.chains.902]
+# ... etc.
+```
+
+Only the minimum necessary state is stored. For example, the rollup config can be derived from deploy config
+variables and therefore can be generated on-the-fly when it is needed. `op-deployer` will expose utility subcommands
+output any generated data, such as the rollup config, to the console.
+
+`op-deployer` itself will be a single Go binary that can be run from the command line. Some stages - like those that
+interact with Forge tooling to deploy contracts - will either shell out to other tools or orchestrate Docker
+containers that run the necessary tools. This way both local development and pre-packaged use cases are supported.
+Given the overall migration towards Go tooling over Forge for genesis creation and deployments, we expect that the
+number of tools requiring Docker/shell integrations will decrease over time.
+
+Each stage will be represented as a standalone Go function for easy integration with additional tooling.
+
+## Deploy Config Cleanup
+
+The deployment config file is a hodgepodge of almost 100 variables which all need to be set in order to deploy a chain.
+Some of these variables are only used in specific contexts, but all of them need to be set in order for some of our
+tools to work. Additionally, the deploy config expects to deploy a single L2, which makes it impossible to deploy
+multiple chains.
+
+To fix this, we will:
+
+- Remove the L1 genesis variables from the deploy config. Deploying a new l1 should be a separate process from
+  deploying an L2.
+- Remove legacy config variables that are no longer used, like `l1UseClique` or `deploymentWaitConfirmations`.
+- Split the deploy config into multiple stanzas, so that irrelevant config variables can be ignored. For example,
+  the DA configuration must be specified right now for the config to validate, even when it's not used by a given chain.
+- Remove booleans that are used to enable/disable features. This data belongs inside of the deployment intent, not
```

Comment:
Ah this is interesting - note that this will require a decent refactor to the way the smart contracts work. It sounds like we will need to modularize some functionality so that it can be composed together and have different entrypoints rather than 1 entrypoint with a config with branching logic

Reply:
> modularize some functionality so that it can be composed together and have different entrypoints rather than 1 entrypoint with a config with branching logic

Yes 100%. I'm happy to help with this as part of this project.

Reply:
See all uses of `cfg.useInterop` in:
- https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/scripts/L2Genesis.s.sol
- https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/scripts/deploy/Deploy.s.sol

Reply:
I like the idea of making the config / intent modular, rather than putting feature-booleans in a huge config